File: src/lib/map/geojson.ts
import type { VectorFeature, VectorLayerData } from "./types.ts";
import { PREVIEW_POLICY, previewNote } from "./preview.ts";
import { crsFromEpsg, resolveGeojsonCrs, type CrsInfo } from "./crs.ts";

function asCoord(pair: unknown): { x: number; y: number } | null {
  if (!Array.isArray(pair) || pair.length < 2) return null;
  const x = Number(pair[0]);
  const y = Number(pair[1]);
  if (!Number.isFinite(x) || !Number.isFinite(y)) return null;
  return { x, y };
}

function lineCoords(coords: unknown): { x: number; y: number }[] {
  if (!Array.isArray(coords)) return [];
  const pts: { x: number; y: number }[] = [];
  for (const item of coords) {
    const p = asCoord(item);
    if (p) pts.push(p);
  }
  return pts;
}

function featuresFromGeometry(
  geom: { type?: string; coordinates?: unknown },
  extras?: { id?: unknown; properties?: Record<string, unknown> }
): VectorFeature[] {
  const type = geom?.type;
  const coordinates = geom?.coordinates;
  const meta = {
    id: extras?.id as string | number | undefined,
    properties: extras?.properties,
  };
  if (type === "Point") {
    const p = asCoord(coordinates);
    return p ? [{ type: "Point", coordinates: [p], ...meta }] : [];
  }
  if (type === "MultiPoint" && Array.isArray(coordinates)) {
    return coordinates
      .map((item) => asCoord(item))
      .filter((p): p is { x: number; y: number } => Boolean(p))
      .map((p) => ({ type: "Point" as const, coordinates: [p], ...meta }));
  }
  if (type === "LineString") {
    const pts = lineCoords(coordinates);
    return pts.length >= 2 ? [{ type: "LineString", coordinates: pts, ...meta }] : [];
  }
  if (type === "MultiLineString" && Array.isArray(coordinates)) {
    return coordinates
      .map((line) => lineCoords(line))
      .filter((pts) => pts.length >= 2)
      .map((pts) => ({ type: "LineString" as const, coordinates: pts, ...meta }));
  }
  if (type === "Polygon" && Array.isArray(coordinates)) {
    const ring = lineCoords(coordinates[0]);
    return ring.length >= 3 ? [{ type: "Polygon", coordinates: ring, ...meta }] : [];
  }
  if (type === "MultiPolygon" && Array.isArray(coordinates)) {
    const out: VectorFeature[] = [];
    for (const poly of coordinates) {
      if (!Array.isArray(poly)) continue;
      const ring = lineCoords(poly[0]);
      if (ring.length >= 3) out.push({ type: "Polygon", coordinates: ring, ...meta });
    }
    return out;
  }
  return [];
}

export function parseGeojson(text: string): { data: VectorLayerData; crs: CrsInfo } | null {
  if (text.length > PREVIEW_POLICY.maxGeojsonBytes) {
    return {
      crs: crsFromEpsg(undefined, "geojson"),
      data: {
        features: [],
        featureCount: 0,
        preview: true,
        previewNote: previewNote("byte-limit"),
      },
    };
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch {
    return null;
  }
  if (!parsed || typeof parsed !== "object") return null;
  const obj = parsed as {
    type?: string;
    features?: { geometry?: { type?: string; coordinates?: unknown }; id?: unknown; properties?: Record<string, unknown> }[];
    geometry?: { type?: string; coordinates?: unknown };
    id?: unknown;
    properties?: Record<string, unknown>;
    coordinates?: unknown;
  };
  const collected: VectorFeature[] = [];
  if (obj.type === "FeatureCollection" && Array.isArray(obj.features)) {
    for (const feature of obj.features) {
      if (feature?.geometry) collected.push(...featuresFromGeometry(feature.geometry, { id: feature.id, properties: feature.properties }));
    }
  } else if (obj.type === "Feature" && obj.geometry) {
    collected.push(...featuresFromGeometry(obj.geometry, { id: obj.id, properties: obj.properties }));
  } else if (obj.type && obj.coordinates) {
    collected.push(...featuresFromGeometry(obj));
  } else {
    return null;
  }
  let bbox: { minX: number; minY: number; maxX: number; maxY: number } | undefined;
  for (const feature of collected) {
    for (const p of feature.coordinates) {
      if (!bbox) bbox = { minX: p.x, minY: p.y, maxX: p.x, maxY: p.y };
      else {
        bbox.minX = Math.min(bbox.minX, p.x);
        bbox.minY = Math.min(bbox.minY, p.y);
        bbox.maxX = Math.max(bbox.maxX, p.x);
        bbox.maxY = Math.max(bbox.maxY, p.y);
      }
    }
  }
  const resolved = resolveGeojsonCrs(parsed, { sourceText: text, bbox });
  const truncated = collected.length > PREVIEW_POLICY.maxGeojsonFeatures;
  return {
    crs: resolved.crs,
    data: {
      features: collected.slice(0, PREVIEW_POLICY.maxGeojsonFeatures),
      featureCount: collected.length,
      preview: truncated,
      previewNote: truncated ? previewNote("truncated-geojson") : undefined,
    },
  };
}

export function pointsFromVector(data: VectorLayerData): { x: number; y: number }[] {
  const pts: { x: number; y: number }[] = [];
  for (const feature of data.features) {
    if (feature.type === "Point") pts.push(...feature.coordinates);
  }
  return pts;
}

export function linesFromVector(data: VectorLayerData): { x: number; y: number }[][] {
  return data.features.filter((feature) => feature.type === "LineString").map((feature) => feature.coordinates);
}

export function polygonsFromVector(data: VectorLayerData): { x: number; y: number }[][] {
  return data.features.filter((feature) => feature.type === "Polygon").map((feature) => feature.coordinates);
}

File: python/formats/geojson.py
"""Documented GeoJSON reader. RFC 7946 default CRS84 is not assumed."""

from __future__ import annotations

import json
import os
import re
from typing import Any

EPSG_RE = re.compile(r"EPSG[:\s]*([0-9]{4,6})", re.I)
AUTHORITY_RE = re.compile(r'AUTHORITY\["EPSG","(\d+)"\]', re.I)
COMMENT_RE = re.compile(r"/\s*EPSG\s*=\s*(\d{4,6})", re.I)


def _finite_pair(value: Any) -> tuple[float, float] | None:
    if not isinstance(value, (list, tuple)) or len(value) < 2:
        return None
    try:
        x = float(value[0])
        y = float(value[1])
    except (TypeError, ValueError):
        return None
    if x != x or y != y:  # NaN
        return None
    return x, y


def _line(coords: Any) -> list[tuple[float, float]]:
    if not isinstance(coords, list):
        return []
    out = []
    for item in coords:
        pair = _finite_pair(item)
        if pair:
            out.append(pair)
    return out


def _closed(ring: list[tuple[float, float]]) -> bool:
    return len(ring) >= 4 and ring[0] == ring[-1]


def _epsg_from_obj(obj: dict) -> int | None:
    crs = obj.get("crs") or {}
    props = crs.get("properties") if isinstance(crs, dict) else {}
    name = (props or {}).get("name") if isinstance(props, dict) else None
    if isinstance(name, str):
        match = EPSG_RE.search(name)
        if match:
            return int(match.group(1))
    extra = obj.get("properties") if isinstance(obj.get("properties"), dict) else {}
    for key in ("EPSG", "crs", "CRS"):
        match = EPSG_RE.search(str(extra.get(key) or ""))
        if match:
            return int(match.group(1))
    return None


def _epsg_from_prj(path: str) -> int | None:
    stem, _ = os.path.splitext(path)
    prj = stem + ".prj"
    if not os.path.isfile(prj):
        return None
    try:
        text = open(prj, encoding="utf-8", errors="replace").read()
    except OSError:
        return None
    auths = list(AUTHORITY_RE.finditer(text))
    if auths:
        return int(auths[-1].group(1))
    match = EPSG_RE.search(text)
    return int(match.group(1)) if match else None


def _walk_geom(geom: dict, errors: list[str]) -> tuple[str, list[tuple[float, float]]]:
    if not isinstance(geom, dict) or not geom.get("type"):
        errors.append("Feature is missing a geometry object.")
        return "", []
    gtype = str(geom.get("type"))
    coords = geom.get("coordinates")
    if gtype == "GeometryCollection":
        pts: list[tuple[float, float]] = []
        last = ""
        for child in geom.get("geometries") or []:
            ctype, cpts = _walk_geom(child if isinstance(child, dict) else {}, errors)
            if cpts:
                last = ctype or last
                pts.extend(cpts)
        return last or "GeometryCollection", pts
    if gtype == "Point":
        pair = _finite_pair(coords)
        if not pair:
            errors.append("Point coordinates are missing or not finite.")
            return gtype, []
        return gtype, [pair]
    if gtype == "MultiPoint" and isinstance(coords, list):
        pts = [p for p in (_finite_pair(item) for item in coords) if p]
        if not pts:
            errors.append("MultiPoint has no finite coordinates.")
        return gtype, pts
    if gtype == "LineString":
        pts = _line(coords)
        if len(pts) < 2:
            errors.append("LineString needs at least two finite positions.")
            return gtype, []
        return gtype, pts
    if gtype == "MultiLineString" and isinstance(coords, list):
        pts: list[tuple[float, float]] = []
        for line in coords:
            line_pts = _line(line)
            if len(line_pts) >= 2:
                pts.extend(line_pts)
        if not pts:
            errors.append("MultiLineString has no valid line.")
        return gtype, pts
    if gtype == "Polygon" and isinstance(coords, list):
        ring = _line(coords[0] if coords else [])
        if not _closed(ring):
            errors.append("Polygon exterior ring must be closed with at least four finite positions.")
            return gtype, []
        return gtype, ring
    if gtype == "MultiPolygon" and isinstance(coords, list):
        pts: list[tuple[float, float]] = []
        for poly in coords:
            if not isinstance(poly, list) or not poly:
                continue
            ring = _line(poly[0])
            if _closed(ring):
                pts.extend(ring)
        if not pts:
            errors.append("MultiPolygon has no valid closed exterior ring.")
        return gtype, pts
    errors.append(f"Geometry type {gtype} is not a supported processing geometry.")
    return gtype, []


def parse_geojson(path: str, role: str | None = None, role_reviewed: bool = False) -> dict:
    text = open(path, encoding="utf-8").read()
    try:
        obj = json.loads(text)
    except json.JSONDecodeError as exc:
        raise ValueError(f"GeoJSON is not valid JSON: {exc}") from exc
    if not isinstance(obj, dict):
        raise ValueError("GeoJSON root must be an object.")

    features_in: list[dict]
    if obj.get("type") == "FeatureCollection" and isinstance(obj.get("features"), list):
        features_in = [item for item in obj["features"] if isinstance(item, dict)]
    elif obj.get("type") == "Feature":
        features_in = [obj]
    elif obj.get("type") and obj.get("coordinates") is not None:
        features_in = [{"type": "Feature", "geometry": obj, "properties": {}}]
    else:
        raise ValueError("Root must be a FeatureCollection, Feature, or geometry object.")

    errors: list[str] = []
    warnings = [
        "Attribute names have unknown semantics. Geology, tenure, and mineral meaning are not inferred from field names or filenames.",
    ]
    features = []
    types: set[str] = set()
    attrs: set[str] = set()
    xs: list[float] = []
    ys: list[float] = []
    for index, feature in enumerate(features_in):
        geom = feature.get("geometry") if isinstance(feature.get("geometry"), dict) else feature
        gtype, pts = _walk_geom(geom if isinstance(geom, dict) else {}, errors)
        if gtype:
            types.add(gtype)
        if not pts:
            continue
        props = feature.get("properties") if isinstance(feature.get("properties"), dict) else {}
        for key in props:
            attrs.add(str(key))
        xs.extend(p[0] for p in pts)
        ys.extend(p[1] for p in pts)
        features.append(
            {
                "id": feature.get("id") if feature.get("id") is not None else index,
                "geometry_type": gtype,
                "coordinates": [{"x": x, "y": y} for x, y in pts],
                "properties": props,
                "semantics": "unknown",
            }
        )

    if not features:
        raise ValueError("No valid geometries after coordinate and ring checks. " + "; ".join(dict.fromkeys(errors)))

    epsg = _epsg_from_obj(obj)
    source = "geojson-crs" if epsg else None
    if epsg is None:
        comment = COMMENT_RE.search(text[:2000]) or EPSG_RE.search(text[:500])
        if comment:
            epsg = int(comment.group(1))
            source = "epsg-comment"
    if epsg is None:
        epsg = _epsg_from_prj(path)
        source = "companion-prj" if epsg else None
    if epsg is None:
        raise ValueError(
            "No documented EPSG. RFC 7946 lon/lat is not assumed. I will not invent a CRS or silently reproject."
        )

    unique_errors = list(dict.fromkeys(errors))
    return {
        "product_name": "G-AID documented GeoJSON vector layer",
        "format": "geojson",
        "crs_epsg": epsg,
        "crs": f"EPSG:{epsg}",
        "crs_source": source,
        "location_quality": "documented",
        "geometry_types": sorted(types),
        "attribute_names": sorted(attrs),
        "feature_count": len(features),
        "features": features,
        "bbox": {
            "minX": min(xs),
            "minY": min(ys),
            "maxX": max(xs),
            "maxY": max(ys),
        },
        "role": role or "generic-vector",
        "role_reviewed": bool(role_reviewed),
        "source_path": path,
        "warnings": warnings,
        "errors": unique_errors,
        "formula": "GeoJSON Feature/FeatureCollection with documented EPSG. Overlay is not geological proof.",
    }
